File: solutions/sonnet-1/sonnet/perl/roman.pl
use strict;
use warnings;

my @values = (1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1);
my @symbols = qw(M CM D CD C XC L XL X IX V IV I);

while (my $line = <STDIN>) {
    chomp $line;
    next unless $line =~ /\S/;
    my $n = $line + 0;
    my $result = '';
    for my $i (0 .. $#values) {
        while ($n >= $values[$i]) {
            $n -= $values[$i];
            $result .= $symbols[$i];
        }
    }
    print "$result\n";
}

File: solutions/sonnet-1/sonnet/ruby/roman.rb
VALUES = [
  [1000, "M"], [900, "CM"], [500, "D"], [400, "CD"],
  [100, "C"], [90, "XC"], [50, "L"], [40, "XL"],
  [10, "X"], [9, "IX"], [5, "V"], [4, "IV"], [1, "I"]
]

def to_roman(n)
  result = ""
  VALUES.each do |value, symbol|
    while n >= value
      result << symbol
      n -= value
    end
  end
  result
end

STDIN.each_line do |line|
  line = line.strip
  next if line.empty?
  puts to_roman(line.to_i)
end
